File: frontend/src/lib/leads/analyticsMetrics.js
/** Метрики воронки и KPI (зеркало backend/core/analytics_metrics.py). */

export const WON_STAGE = 'Выигран';
export const LOST_STAGE = 'Проигран';

export function parseBudgetRub(raw) {
	if (!raw) return null;
	const s = String(raw).replace(/\s/g, '').replace(/[^\d.,]/g, '').replace(',', '.');
	if (!s) return null;
	const v = Number(s);
	return Number.isFinite(v) && v > 0 ? v : null;
}

export function leadAmountRub(lead) {
	if (lead?.amountRub != null && lead.amountRub !== '') {
		const v = Number(lead.amountRub);
		if (Number.isFinite(v) && v > 0) return v;
	}
	return parseBudgetRub(lead?.budget);
}

function parseIsoDate(value) {
	if (!value) return null;
	const d = new Date(value);
	return Number.isNaN(d.getTime()) ? null : d;
}

function parseCreatedDate(lead) {
	if (lead?.createdAt) return parseIsoDate(lead.createdAt);
	const c = lead?.created;
	if (c && c !== '—') {
		const m = /^(\d{2})\.(\d{2})\.(\d{4})$/.exec(String(c));
		if (m) return new Date(Number(m[3]), Number(m[2]) - 1, Number(m[1]));
	}
	return null;
}

function normalizeStage(stage, stages) {
	const s = String(stage || 'Новый').trim();
	return stages.includes(s) ? s : 'Новый';
}

/** @param {object[]} leads @param {string[]} stages */
export function computeAnalytics(leads, stages) {
	const sts = stages?.length ? stages : [];
	const byStage = Object.fromEntries(sts.map((s) => [s, 0]));
	const total = leads.length;
	let won = 0;
	let lost = 0;
	const dealAmounts = [];
	const cycleDays = [];

	for (const lead of leads) {
		const st = normalizeStage(lead.stage, sts);
		byStage[st] = (byStage[st] || 0) + 1;
		if (st === WON_STAGE) {
			won += 1;
			const amt = leadAmountRub(lead);
			if (amt != null) dealAmounts.push(amt);
			const created = parseCreatedDate(lead);
			const updated = parseIsoDate(lead.updatedAt);
			if (created && updated) {
				const delta = Math.round((updated - created) / 86400000);
				if (delta >= 0) cycleDays.push(delta);
			}
		} else if (st === LOST_STAGE) {
			lost += 1;
		}
	}

	const funnel = sts.map((stage) => {
		const count = byStage[stage] || 0;
		const pct = total ? Math.round((1000 * count) / total) / 10 : 0;
		return { stage, count, pct };
	});

	const conversionPct = total ? Math.round((1000 * won) / total) / 10 : 0;
	const winRatePct = won + lost ? Math.round((1000 * won) / (won + lost)) / 10 : null;
	const avgDealRub = dealAmounts.length
		? Math.round(dealAmounts.reduce((a, b) => a + b, 0) / dealAmounts.length)
		: null;
	const avgCycleDays = cycleDays.length
		? Math.round(cycleDays.reduce((a, b) => a + b, 0) / cycleDays.length)
		: null;
	const avgScore = total
		? Math.round(leads.reduce((a, l) => a + (Number(l.score) || 0), 0) / total)
		: 0;

	return {
		total,
		wonCount: won,
		lostCount: lost,
		avgScore,
		byStage,
		funnel,
		conversionPct,
		winRatePct,
		avgDealRub,
		avgCycleDays,
	};
}

export function formatRub(value) {
	if (value == null) return '—';
	return new Intl.NumberFormat('ru-RU', {
		style: 'currency',
		currency: 'RUB',
		maximumFractionDigits: 0,
	}).format(value);
}

File: backend/core/analytics_metrics.py
"""Метрики воронки и KPI из списка лидов (зеркало frontend analyticsMetrics.js)."""
from __future__ import annotations

import csv
import io
import re
from datetime import datetime
from typing import Any

WON_STAGE = "Выигран"
LOST_STAGE = "Проигран"

DEFAULT_STAGES = [
    "Новый",
    "Квалифицирован",
    "КП отправлено",
    "Переговоры",
    "Выигран",
    "Проигран",
]


def parse_budget_rub(raw: str | None) -> float | None:
    if not raw:
        return None
    s = re.sub(r"[^\d.,]", "", str(raw).replace(" ", ""))
    if not s:
        return None
    s = s.replace(",", ".")
    try:
        v = float(s)
        return v if v > 0 else None
    except ValueError:
        return None


def lead_amount_rub(lead: dict[str, Any]) -> float | None:
    ar = lead.get("amountRub")
    if ar is not None and ar != "":
        try:
            v = float(ar)
            return v if v > 0 else None
        except (TypeError, ValueError):
            pass
    return parse_budget_rub(lead.get("budget"))


def parse_iso_date(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        return datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    except ValueError:
        return None


def parse_created_date(lead: dict[str, Any]) -> datetime | None:
    if lead.get("createdAt"):
        return parse_iso_date(str(lead["createdAt"]))
    created = lead.get("created")
    if created and created != "—":
        try:
            return datetime.strptime(str(created), "%d.%m.%Y")
        except ValueError:
            pass
    return None


def normalize_stage(stage: str | None, stages: list[str]) -> str:
    s = (stage or "Новый").strip()
    return s if s in stages else "Новый"


def compute_analytics(leads: list[dict[str, Any]], stages: list[str] | None = None) -> dict[str, Any]:
    sts = stages or DEFAULT_STAGES
    by_stage = {s: 0 for s in sts}
    total = len(leads)
    won = 0
    lost = 0
    deal_amounts: list[float] = []
    cycle_days: list[int] = []

    for lead in leads:
        st = normalize_stage(lead.get("stage"), sts)
        by_stage[st] = by_stage.get(st, 0) + 1
        if st == WON_STAGE:
            won += 1
            amt = lead_amount_rub(lead)
            if amt is not None:
                deal_amounts.append(amt)
            created = parse_created_date(lead)
            updated = parse_iso_date(lead.get("updatedAt"))
            if created and updated:
                delta = (updated - created).days
                if delta >= 0:
                    cycle_days.append(delta)
        elif st == LOST_STAGE:
            lost += 1

    funnel = []
    for stage in sts:
        count = by_stage.get(stage, 0)
        pct = round(100 * count / total, 1) if total else 0.0
        funnel.append({"stage": stage, "count": count, "pct": pct})

    conversion_pct = round(100 * won / total, 1) if total else 0.0
    win_rate_pct = round(100 * won / (won + lost), 1) if (won + lost) else None
    avg_deal_rub = round(sum(deal_amounts) / len(deal_amounts)) if deal_amounts else None
    avg_cycle_days = round(sum(cycle_days) / len(cycle_days)) if cycle_days else None
    avg_score = round(sum(float(l.get("score") or 0) for l in leads) / total) if total else 0

    return {
        "total": total,
        "wonCount": won,
        "lostCount": lost,
        "avgScore": avg_score,
        "byStage": by_stage,
        "funnel": funnel,
        "conversionPct": conversion_pct,
        "winRatePct": win_rate_pct,
        "avgDealRub": avg_deal_rub,
        "avgCycleDays": avg_cycle_days,
    }


def analytics_to_csv(metrics: dict[str, Any]) -> str:
    buf = io.StringIO()
    w = csv.writer(buf, lineterminator="\n")
    w.writerow(["metric", "value"])
    w.writerow(["total", metrics["total"]])
    w.writerow(["conversionPct", metrics["conversionPct"]])
    if metrics.get("winRatePct") is not None:
        w.writerow(["winRatePct", metrics["winRatePct"]])
    w.writerow(["avgDealRub", metrics.get("avgDealRub") or ""])
    w.writerow(["avgCycleDays", metrics.get("avgCycleDays") or ""])
    w.writerow(["avgScore", metrics.get("avgScore", 0)])
    w.writerow([])
    w.writerow(["stage", "count", "pct"])
    for row in metrics["funnel"]:
        w.writerow([row["stage"], row["count"], row["pct"]])
    return buf.getvalue()
